Automattic/jetpack
# Stats: allow filtering the list of attributes used in script tag

Fixes #29560

## Proposed changes:

This is a bit involved; I could have gone with a much simpler approach with a filterable string that would have been just `defer` by default.

I went with a more complex approach with an array, for 2 reasons:

1. It may be easier for third-parties to add and remove from an array based on a specific condition, instead of having to append to a string or check for the string contents and add / remove from it.
2. It allows us to escape each key / value before we output it.
3. It allows us to ensure that we output valid html.

To use the filter, you would add the following to a functionality plugin:

```php
/**
 * Add custom attribute to the Stats tracking script.
 *
 * @param array $script_attributes Array of attributes to add to the script tag.
 */
function jeherve_add_stats_script_attributes( $script_attributes ) {
	$script_attributes['type'] = 'text/plain';
	$script_attributes['async'] = true;
	$script_attributes['thing'] = false;
	$script_attributes['what’now'] = '<script>alert("hello");</script>';
	return $script_attributes;
}
add_filter( 'jetpack_stats_footer_script_attributes', 'jeherve_add_stats_script_attributes' );
```

### Other information:


## Jetpack product discussion

* No

## Does this pull request change what data or activity we track or use?

* No

## Testing instructions:

1. Start with a site that's connected to WordPress.com, with the stats module active.
2. Open the site in an incognito window, and look for the `<script src='https://stats.wp.com/` script tag.
    * It should include just the `defer` attribute.
3. Now add the code snippet above to a functionality plugin.
    * Ensure the new attributes are correctly added when necessary.
4. Do the tests pass?
